File: Packages/LookAfterCore/Sources/LookAfterCore/Execution/ExecutionBlockSnapshot.swift
import Foundation

/// Visual / behavioral mode for Live Activity and Focus surfaces.
public enum ExecutionSurfaceMode: String, Codable, Sendable, Hashable {
    /// High-contrast timer — immovable focus block.
    case anchored
    /// Preferred slot with confidence — active focus allowed.
    case flexible
    /// Soft ambient indicator — no countdown pressure.
    case recovery
    /// Minimal unstructured gap before the next commitment.
    case fluidGap
    /// Nothing to project (day ended / empty schedule).
    case idle

    public var displayName: String {
        switch self {
        case .anchored: return "Anchored Window"
        case .flexible: return "Flexible Focus"
        case .recovery: return "Recovery Window"
        case .fluidGap: return "Fluid Gap"
        case .idle: return "Idle"
        }
    }

    /// Whether a Live Activity should remain visible for this mode.
    public var projectsLiveActivity: Bool {
        self != .idle
    }

    /// Whether the Dynamic Island should show a strict countdown timer.
    public var showsStrictCountdown: Bool {
        switch self {
        case .anchored, .flexible:
            return true
        case .recovery, .fluidGap, .idle:
            return false
        }
    }
}

/// Immutable projection of the currently active execution block.
/// Produced by `ExecutionBlockResolver` — UI / ActivityKit only consume this.
public struct ExecutionBlockSnapshot: Codable, Sendable, Equatable, Identifiable {
    public var id: String
    public var taskID: String?
    public var taskTitle: String
    public var category: FocusTaskCategory
    public var surfaceMode: ExecutionSurfaceMode
    public var constraintType: TimeConstraint?
    public var windowStart: Date
    public var windowEnd: Date
    public var progressFraction: Double
    public var nextUpSummary: String
    public var confidence: Double
    public var generatedAt: Date

    public init(
        id: String = UUID().uuidString,
        taskID: String? = nil,
        taskTitle: String,
        category: FocusTaskCategory,
        surfaceMode: ExecutionSurfaceMode,
        constraintType: TimeConstraint? = nil,
        windowStart: Date,
        windowEnd: Date,
        progressFraction: Double,
        nextUpSummary: String = "",
        confidence: Double = 1.0,
        generatedAt: Date = Date()
    ) {
        self.id = id
        self.taskID = taskID
        self.taskTitle = taskTitle
        self.category = category
        self.surfaceMode = surfaceMode
        self.constraintType = constraintType
        self.windowStart = windowStart
        self.windowEnd = windowEnd
        self.progressFraction = min(1, max(0, progressFraction))
        self.nextUpSummary = nextUpSummary
        self.confidence = min(1, max(0, confidence))
        self.generatedAt = generatedAt
    }

    public var remainingSeconds: TimeInterval {
        max(0, windowEnd.timeIntervalSince(generatedAt))
    }

    public var isFocusEligible: Bool {
        switch surfaceMode {
        case .anchored, .flexible:
            return true
        case .recovery, .fluidGap, .idle:
            return false
        }
    }

    /// Constraint label projected onto Live Activity content state.
    public var constraintLabel: String {
        switch surfaceMode {
        case .anchored: return "Anchored"
        case .flexible: return "Flexible"
        case .recovery: return "Recovery"
        case .fluidGap: return "Fluid"
        case .idle: return "Idle"
        }
    }

    public static let idle = ExecutionBlockSnapshot(
        id: "idle",
        taskTitle: "",
        category: .fluidGap,
        surfaceMode: .idle,
        windowStart: .distantPast,
        windowEnd: .distantPast,
        progressFraction: 0,
        nextUpSummary: "",
        confidence: 0
    )
}

File: android/lookafter-core/src/main/kotlin/com/lookafter/core/models/ExecutionBlockSnapshot.kt
package com.lookafter.core.models

import java.time.Duration
import java.time.Instant
import java.util.UUID
import kotlin.math.max
import kotlin.math.min

/**
 * Immutable projection of the currently active execution block.
 * Produced by execution resolvers — UI only consumes this.
 * Mirrors iOS `ExecutionBlockSnapshot`.
 */
data class ExecutionBlockSnapshot(
    val id: String = UUID.randomUUID().toString(),
    val taskId: String? = null,
    val taskTitle: String,
    val category: FocusTaskCategory,
    val surfaceMode: ExecutionSurfaceMode,
    val constraintType: ConstraintType? = null,
    val windowStart: Instant,
    val windowEnd: Instant,
    val progressFraction: Double,
    val nextUpSummary: String = "",
    val confidence: Double = 1.0,
    val generatedAt: Instant = Instant.now(),
) {
    val remainingSeconds: Long
        get() = max(0L, Duration.between(generatedAt, windowEnd).seconds)

    val isFocusEligible: Boolean
        get() = when (surfaceMode) {
            ExecutionSurfaceMode.ANCHORED, ExecutionSurfaceMode.FLEXIBLE -> true
            ExecutionSurfaceMode.RECOVERY,
            ExecutionSurfaceMode.FLUID_GAP,
            ExecutionSurfaceMode.IDLE,
            -> false
        }

    val constraintLabel: String
        get() = when (surfaceMode) {
            ExecutionSurfaceMode.ANCHORED -> "Anchored"
            ExecutionSurfaceMode.FLEXIBLE -> "Flexible"
            ExecutionSurfaceMode.RECOVERY -> "Recovery"
            ExecutionSurfaceMode.FLUID_GAP -> "Fluid"
            ExecutionSurfaceMode.IDLE -> "Idle"
        }

    companion object {
        val IDLE: ExecutionBlockSnapshot = ExecutionBlockSnapshot(
            id = "idle",
            taskTitle = "",
            category = FocusTaskCategory.FLUID_GAP,
            surfaceMode = ExecutionSurfaceMode.IDLE,
            windowStart = Instant.EPOCH,
            windowEnd = Instant.EPOCH,
            progressFraction = 0.0,
            confidence = 0.0,
            generatedAt = Instant.EPOCH,
        )

        fun clamped(
            id: String = UUID.randomUUID().toString(),
            taskId: String? = null,
            taskTitle: String,
            category: FocusTaskCategory,
            surfaceMode: ExecutionSurfaceMode,
            constraintType: ConstraintType? = null,
            windowStart: Instant,
            windowEnd: Instant,
            progressFraction: Double,
            nextUpSummary: String = "",
            confidence: Double = 1.0,
            generatedAt: Instant = Instant.now(),
        ): ExecutionBlockSnapshot = ExecutionBlockSnapshot(
            id = id,
            taskId = taskId,
            taskTitle = taskTitle,
            category = category,
            surfaceMode = surfaceMode,
            constraintType = constraintType,
            windowStart = windowStart,
            windowEnd = windowEnd,
            progressFraction = min(1.0, max(0.0, progressFraction)),
            nextUpSummary = nextUpSummary,
            confidence = min(1.0, max(0.0, confidence)),
            generatedAt = generatedAt,
        )
    }
}
